packages/e2e/lib/common.dart

```diff
@@ -0,0 +1,57 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+
+/// An object sent from e2e back to the Flutter Driver in response to
+/// `request_data` command.
+class Response {
+  String _failureDetails;
+
+  final bool _allTestsPassed;
+
+  /// Constructor which receives tests results' flag and failure details.
+  Response(this._allTestsPassed, String failureDetails) {
+    this._failureDetails = failureDetails;
+  }
+
+  /// Whether the test run successfully or not.
+  String get result => _allTestsPassed ? 'pass' : 'fail';
+
+  /// If the result are failures get the formatted details.
+  String get failureDetails => _allTestsPassed ? '' : _failureDetails;
+
+  /// Convert a string pass/fail result to a boolean.
+  static bool allTestsPassed(String result) => (result == 'pass')
+      ? true
+      : (result == 'fail') ? false : throw 'Invalid State for result.';
+
+  /// Serializes this message to a JSON map.
+  String toJson() => json.encode(<String, String>{
+        'result': result,
+        'failureDetails': _failureDetails,
+      });
+
+  /// Deserializes the result from JSON.
+  static Response fromJson(String source) {
+    Map<String, dynamic> result = json.decode(source);
+    return Response(allTestsPassed(result['result']), result['failureDetails']);
+  }
+}
+
+/// Method for formating the test failures' details.
+String formatFailures(Map<String, String> failureDetails) {
```

Comment:
If failure details are a `Map<String, String>` then it can be turned to JSON as is. Why format? Formatting loses information that could be used on the receiving end.

Reply:
Cause formatting looks better on the output. It adds extra empty lines for example.

For me it increased my productivity since I am able to see things a lot more clear. You don't have to use it in your tests though. This one is not a mandatory method.

Reply:
For giving more context this is as json:
`{"verify text":"══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════\nThe following TestFailure object was thrown running a test:\n  Expected: <5>\n  Actual: <4>\n\nWhen the exception was thrown, this was the stack:\n    at Object.wrapException (http://localhost:8080/main.dart.js:3646:17)\n    at Object.throwExpression (http://localhost:8080/main.dart.js:3659:15)\n    at Object.fail (http://localhost:8080/main.dart.js:16444:16)\n    at Object._expect (http://localhost:8080/main.dart.js:16441:9)\n    at Object.expect0 (http://localhost:8080/main.dart.js:16417:9)\n    at Object.expect (http://localhost:8080/main.dart.js:19769:9)\n    at http://localhost:8080/main.dart.js:82327:17\n    at _wrapJsFunctionForAsync_closure.$protected (http://localhost:8080/main.dart.js:6615:15)\n    at _wrapJsFunctionForAsync_closure.call$2 (http://localhost:8080/main.dart.js:37092:12)\n    at StackZoneSpecification__registerBinaryCallback__closure.call$0 (http://localhost:8080/main.dart.js:80369:21)\n\nThe test description was:\n  verify text\n═════════════════════════════════════════════════════════════════\n","fail text":"══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════\nThe following TestFailure object was thrown running a test:\n  Expected: <5>\n  Actual: <4>\n\nWhen the exception was thrown, this was the stack:\n    at Object.wrapException (http://localhost:8080/main.dart.js:3646:17)\n    at Object.throwExpression (http://localhost:8080/main.dart.js:3659:15)\n    at Object.fail (http://localhost:8080/main.dart.js:16444:16)\n    at Object._expect (http://localhost:8080/main.dart.js:16441:9)\n    at Object.expect0 (http://localhost:8080/main.dart.js:16417:9)\n    at Object.expect (http://localhost:8080/main.dart.js:19769:9)\n    at http://localhost:8080/main.dart.js:82349:17\n    at _wrapJsFunctionForAsync_closure.$protected (http://localhost:8080/main.dart.js:6615:15)\n    at _wrapJsFunctionForAsync_closure.call$2 (http://localhost:8080/main.dart.js:37092:12)\n    at StackZoneSpecification__registerBinaryCallback__closure.call$0 (http://localhost:8080/main.dart.js:80369:21)\n\nThe test description was:\n  fail text\n═════════════════════════════════════════════════════════════════\n"}
`

this is with my format method:
`Failure Details:
Failure in method: verify text
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following TestFailure object was thrown running a test:
  Expected: <5>
  Actual: <4>

When the exception was thrown, this was the stack:
    at Object.wrapException (http://localhost:8080/main.dart.js:3646:17)
    at Object.throwExpression (http://localhost:8080/main.dart.js:3659:15)
    at Object.fail (http://localhost:8080/main.dart.js:16453:16)
    at Object._expect (http://localhost:8080/main.dart.js:16450:9)
    at Object.expect0 (http://localhost:8080/main.dart.js:16426:9)
    at Object.expect (http://localhost:8080/main.dart.js:19778:9)
    at http://localhost:8080/main.dart.js:82348:17
    at _wrapJsFunctionForAsync_closure.$protected (http://localhost:8080/main.dart.js:6615:15)
    at _wrapJsFunctionForAsync_closure.call$2 (http://localhost:8080/main.dart.js:37101:12)
    at StackZoneSpecification__registerBinaryCallback__closure.call$0 (http://localhost:8080/main.dart.js:80390:21)

The test description was:
  verify text
═════════════════════════════════════════════════════════════════

end of failure 1


Failure in method: fail text
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞═════════════════
The following TestFailure object was thrown running a test:
  Expected: <5>
  Actual: <4>

When the exception was thrown, this was the stack:
    at Object.wrapException (http://localhost:8080/main.dart.js:3646:17)
    at Object.throwExpression (http://localhost:8080/main.dart.js:3659:15)
    at Object.fail (http://localhost:8080/main.dart.js:16453:16)
    at Object._expect (http://localhost:8080/main.dart.js:16450:9)
    at Object.expect0 (http://localhost:8080/main.dart.js:16426:9)
    at Object.expect (http://localhost:8080/main.dart.js:19778:9)
    at http://localhost:8080/main.dart.js:82370:17
    at _wrapJsFunctionForAsync_closure.$protected (http://localhost:8080/main.dart.js:6615:15)
    at _wrapJsFunctionForAsync_closure.call$2 (http://localhost:8080/main.dart.js:37101:12)
    at StackZoneSpecification__registerBinaryCallback__closure.call$0 (http://localhost:8080/main.dart.js:80390:21)

The test description was:
  fail text
═════════════════════════════════════════════════════════════════

end of failure 2



`

I think I prefer to use the second one. It is more readable hence increases productivity.

Reply:
Ok, I stopped using the formatted result. I added a method which developers can use for a formatted output though.

Reply:
I also added it to our tests:
```
  final common.Response response = common.Response.fromJson(jsonResult);
  await driver.close();

  if (response.allTestsPassed) {
    print('All tests passed.');
    exit(0);
  } else {
    response.formattedFailureDetails;
    print('Failure Details:\n${response.formattedFailureDetails}');
    exit(1);
  }
```